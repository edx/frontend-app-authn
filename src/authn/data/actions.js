import { AsyncActionType } from '../../data/utils';

export const REGISTER_NEW_USER = new AsyncActionType('REGISTRATION', 'REGISTER_NEW_USER');
export const LOGIN_REQUEST = new AsyncActionType('LOGIN', 'REQUEST');
export const THIRD_PARTY_AUTH_CONTEXT = new AsyncActionType('THIRD_PARTY_AUTH', 'GET_THIRD_PARTY_AUTH_CONTEXT');
export const REGISTER_FORM = new AsyncActionType('REGISTRATION', 'GET_FORM_FIELDS');
export const REGISTER_FORM_VALIDATIONS = new AsyncActionType('REGISTRATION', 'GET_FORM_VALIDATIONS');

// Register

export const registerNewUser = registrationInfo => ({
  type: REGISTER_NEW_USER.BASE,
  payload: { registrationInfo },
});

export const registerNewUserBegin = () => ({
  type: REGISTER_NEW_USER.BEGIN,
});

export const registerNewUserSuccess = (redirectUrl, success) => ({
  type: REGISTER_NEW_USER.SUCCESS,
  payload: { redirectUrl, success },

});

export const registerNewUserFailure = (error) => ({
  type: REGISTER_NEW_USER.FAILURE,
  payload: { error },
});

// Login
export const loginRequest = creds => ({
  type: LOGIN_REQUEST.BASE,
  payload: { creds },
});

export const loginRequestBegin = () => ({
  type: LOGIN_REQUEST.BEGIN,
});

export const loginRequestSuccess = (redirectUrl, success) => ({
  type: LOGIN_REQUEST.SUCCESS,
  payload: { redirectUrl, success },
});

export const loginRequestFailure = (loginError) => ({
  type: LOGIN_REQUEST.FAILURE,
  payload: { loginError },
});

// Third party auth context
export const getThirdPartyAuthContext = (urlParams) => ({
  type: THIRD_PARTY_AUTH_CONTEXT.BASE,
  payload: { urlParams },
});

export const getThirdPartyAuthContextBegin = () => ({
  type: THIRD_PARTY_AUTH_CONTEXT.BEGIN,
});

export const getThirdPartyAuthContextSuccess = (thirdPartyAuthContext) => ({
  type: THIRD_PARTY_AUTH_CONTEXT.SUCCESS,
  payload: { thirdPartyAuthContext },
});

export const getThirdPartyAuthContextFailure = () => ({
  type: THIRD_PARTY_AUTH_CONTEXT.FAILURE,
});

// Registration Form Fields
export const fetchRegistrationForm = () => ({
  type: REGISTER_FORM.BASE,
});

export const fetchRegistrationFormBegin = () => ({
  type: REGISTER_FORM.BEGIN,
});

export const fetchRegistrationFormSuccess = (formData) => ({
  type: REGISTER_FORM.SUCCESS,
  payload: { formData },
});

export const fetchRegistrationFormFailure = () => ({
  type: REGISTER_FORM.FAILURE,
});

// Realtime Field validations
export const fetchRealtimeValidations = (formPayload) => ({
  type: REGISTER_FORM_VALIDATIONS.BASE,
  payload: { formPayload },
});

export const fetchRealtimeValidationsBegin = () => ({
  type: REGISTER_FORM_VALIDATIONS.BEGIN,
});

export const fetchRealtimeValidationsSuccess = (validations) => ({
  type: REGISTER_FORM_VALIDATIONS.SUCCESS,
  payload: { validations },
});

export const fetchRealtimeValidationsFailure = (error, statusCode) => ({
  type: REGISTER_FORM_VALIDATIONS.FAILURE,
  payload: { error, statusCode },
});
