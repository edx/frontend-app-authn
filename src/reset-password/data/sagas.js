import { call, put, takeEvery } from 'redux-saga/effects';
import { logError } from '@edx/frontend-platform/logging';

// Actions
import {
  VALIDATE_TOKEN,
  validateTokenBegin,
  validateTokenSuccess,
  validateTokenFailure,
  RESET_PASSWORD,
  resetPasswordBegin,
  resetPasswordSuccess,
  resetPasswordFailure,
} from './actions';

import { validateToken, resetPassword } from './service';

// Services
export function* handleValidateToken(action) {
  try {
    yield put(validateTokenBegin());
    const data = yield call(validateToken, action.payload.token);
    const isValid = data.is_valid;
    if (isValid) {
      yield put(validateTokenSuccess(isValid, action.payload.token));
    } else {
      yield put(validateTokenFailure(isValid));
    }
  } catch (err) {
    yield put(validateTokenFailure(err));
    logError(err);
  }
}

export function* handleResetPassword(action) {
  try {
    yield put(resetPasswordBegin());
    const data = yield call(resetPassword, action.payload.formPayload, action.payload.token, action.payload.params);
    const resetStatus = data.reset_status;
    const resetErrors = data.err_msg;

    if (resetStatus) {
      yield put(resetPasswordSuccess(resetStatus));
    } else {
      yield put(resetPasswordFailure(resetErrors));
    }
  } catch (err) {
    yield put(resetPasswordFailure(err));
    logError(err);
  }
}

export default function* saga() {
  yield takeEvery(RESET_PASSWORD.BASE, handleResetPassword);
  yield takeEvery(VALIDATE_TOKEN.BASE, handleValidateToken);
}
