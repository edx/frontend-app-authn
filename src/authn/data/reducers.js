import {
  REGISTER_NEW_USER,
  LOGIN_REQUEST,
  THIRD_PARTY_AUTH_CONTEXT,
  REGISTER_FORM,
  REGISTER_FORM_VALIDATIONS,
} from './actions';

import { DEFAULT_STATE, PENDING_STATE, COMPLETE_STATE } from '../../data/constants';

export const defaultState = {
  loginError: null,
  loginResult: {},
  registrationError: null,
  registrationResult: {},
  formData: null,
  validations: null,
  statusCode: null,
};

const reducer = (state = defaultState, action) => {
  switch (action.type) {
    case REGISTER_NEW_USER.BEGIN:
      return {
        ...state,
        submitState: PENDING_STATE,
      };
    case REGISTER_NEW_USER.SUCCESS:
      return {
        ...state,
        registrationResult: action.payload,
      };
    case REGISTER_NEW_USER.FAILURE:
      return {
        ...state,
        registrationError: action.payload.error,
        submitState: DEFAULT_STATE,
      };
    case LOGIN_REQUEST.BEGIN:
      return {
        ...state,
        submitState: PENDING_STATE,
      };
    case LOGIN_REQUEST.SUCCESS:
      return {
        ...state,
        loginResult: action.payload,
      };
    case LOGIN_REQUEST.FAILURE:
      return {
        ...state,
        loginError: action.payload.loginError,
        submitState: DEFAULT_STATE,
      };
    case THIRD_PARTY_AUTH_CONTEXT.BEGIN:
      return {
        ...state,
        thirdPartyAuthApiStatus: PENDING_STATE,
      };
    case THIRD_PARTY_AUTH_CONTEXT.SUCCESS:
      return {
        ...state,
        thirdPartyAuthContext: action.payload.thirdPartyAuthContext,
        thirdPartyAuthApiStatus: COMPLETE_STATE,
      };
    case THIRD_PARTY_AUTH_CONTEXT.FAILURE:
      return {
        ...state,
        thirdPartyAuthApiStatus: COMPLETE_STATE,
      };
    case REGISTER_FORM.BEGIN:
      return {
        ...state,
      };
    case REGISTER_FORM.SUCCESS:
      return {
        ...state,
        formData: action.payload.formData,
      };
    case REGISTER_FORM.FAILURE:
      return {
        ...state,
      };
    case REGISTER_FORM_VALIDATIONS.BEGIN:
      return {
        ...state,
      };
    case REGISTER_FORM_VALIDATIONS.SUCCESS:
      return {
        ...state,
        validations: action.payload.validations,
      };
    case REGISTER_FORM_VALIDATIONS.FAILURE:
      return {
        ...state,
        validations: action.payload.error,
        statusCode: action.payload.statusCode,
      };
    default:
      return state;
  }
};

export default reducer;
