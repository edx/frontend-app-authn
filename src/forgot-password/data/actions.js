import { AsyncActionType } from '../../data/utils';

export const FORGOT_PASSWORD = new AsyncActionType('FORGOT', 'PASSWORD');

// Forgot Password
export const forgotPassword = email => ({
  type: FORGOT_PASSWORD.BASE,
  payload: { email },
});

export const forgotPasswordBegin = () => ({
  type: FORGOT_PASSWORD.BEGIN,
});

export const forgotPasswordSuccess = email => ({
  type: FORGOT_PASSWORD.SUCCESS,
  payload: { email },
});

export const forgotPasswordForbidden = (errorCode) => ({
  type: FORGOT_PASSWORD.FORBIDDEN,
  payload: { errorCode },
});
