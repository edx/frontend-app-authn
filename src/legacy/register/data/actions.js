import { AsyncActionType } from '../../data/utils';

export const REGISTER_NEW_USER = new AsyncActionType('REGISTRATION', 'REGISTER_NEW_USER');
export const REGISTER_FORM_VALIDATIONS = new AsyncActionType('REGISTRATION', 'GET_FORM_VALIDATIONS');

// Register
export const registerNewUser = registrationInfo => ({
  type: REGISTER_NEW_USER.BASE,
  payload: { registrationInfo },
});

export const registerNewUserBegin = () => ({
  type: REGISTER_NEW_USER.BEGIN,
});

export const registerNewUserSuccess = (redirectUrl, success) => ({
  type: REGISTER_NEW_USER.SUCCESS,
  payload: { redirectUrl, success },

});

export const registerNewUserFailure = (error) => ({
  type: REGISTER_NEW_USER.FAILURE,
  payload: { error },
});

// Realtime Field validations
export const fetchRealtimeValidations = (formPayload) => ({
  type: REGISTER_FORM_VALIDATIONS.BASE,
  payload: { formPayload },
});

export const fetchRealtimeValidationsBegin = () => ({
  type: REGISTER_FORM_VALIDATIONS.BEGIN,
});

export const fetchRealtimeValidationsSuccess = (validations) => ({
  type: REGISTER_FORM_VALIDATIONS.SUCCESS,
  payload: { validations },
});

export const fetchRealtimeValidationsFailure = (error, statusCode) => ({
  type: REGISTER_FORM_VALIDATIONS.FAILURE,
  payload: { error, statusCode },
});
