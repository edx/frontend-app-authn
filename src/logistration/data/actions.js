import { AsyncActionType } from '../../data/utils';

export const REGISTER_NEW_USER = new AsyncActionType('REGISTRATION', 'REGISTER_NEW_USER');
export const LOGIN_REQUEST = new AsyncActionType('LOGIN', 'REQUEST');
export const THIRD_PARTY_AUTH_CONTEXT = new AsyncActionType('THIRD_PARTY_AUTH', 'GET_THIRD_PARTY_AUTH_CONTEXT');

// Register

export const registerNewUser = registrationInfo => ({
  type: REGISTER_NEW_USER.BASE,
  payload: { registrationInfo },
});

export const registerNewUserBegin = () => ({
  type: REGISTER_NEW_USER.BEGIN,
});

export const registerNewUserSuccess = (redirectUrl, success) => ({
  type: REGISTER_NEW_USER.SUCCESS,
  payload: { redirectUrl, success },
});

export const registerNewUserFailure = () => ({
  type: REGISTER_NEW_USER.FAILURE,
});

// Login
export const loginRequest = creds => ({
  type: LOGIN_REQUEST.BASE,
  payload: { creds },
});

export const loginRequestBegin = () => ({
  type: LOGIN_REQUEST.BEGIN,
});

export const loginRequestSuccess = (redirectUrl, success) => ({
  type: LOGIN_REQUEST.SUCCESS,
  payload: { redirectUrl, success },
});

export const loginRequestFailure = () => ({
  type: LOGIN_REQUEST.FAILURE,
});

// Third party auth context
export const getThirdPartyAuthContext = (urlParams) => ({
  type: THIRD_PARTY_AUTH_CONTEXT.BASE,
  payload: { urlParams },
});

export const getThirdPartyAuthContextBegin = () => ({
  type: THIRD_PARTY_AUTH_CONTEXT.BEGIN,
});

export const getThirdPartyAuthContextSuccess = (thirdPartyAuthContext) => ({
  type: THIRD_PARTY_AUTH_CONTEXT.SUCCESS,
  payload: { thirdPartyAuthContext },
});

export const getThirdPartyAuthContextFailure = () => ({
  type: THIRD_PARTY_AUTH_CONTEXT.FAILURE,
});
