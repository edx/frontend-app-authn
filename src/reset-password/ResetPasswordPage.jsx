import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  Alert, Form, Input, StatefulButton, ValidationFormGroup,
} from '@edx/paragon';
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import { getQueryParameters } from '@edx/frontend-platform';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';

import messages from './messages';
import { resetPassword, validateToken } from './data/actions';
import { resetPasswordResultSelector } from './data/selectors';
import { validatePassword } from './data/service';
import InvalidTokenMessage from './InvalidToken';
import ResetSuccessMessage from './ResetSuccess';
import Spinner from './Spinner';

const ResetPasswordPage = (props) => {
  const { intl } = props;
  const params = getQueryParameters();
  const mainRef = useRef();

  const [newPasswordInput, setNewPasswordValue] = useState('');
  const [confirmPasswordInput, setConfirmPasswordValue] = useState('');
  const [passwordValid, setPasswordValidValue] = useState(true);
  const [passwordMatch, setPasswordMatchValue] = useState(true);
  const [validationMessage, setvalidationMessage] = useState('');
  const [emptyFieldError, setEmptyFieldError] = useState('');

  const validatePasswordFromBackend = async (newPassword) => {
    let errorMessage;
    try {
      errorMessage = await validatePassword(newPassword);
    } catch (err) {
      errorMessage = '';
    }
    setPasswordValidValue(!errorMessage);
    setvalidationMessage(errorMessage);
  };

  const handleNewPasswordChange = (e) => {
    const newPassword = e.target.value;
    setNewPasswordValue(newPassword);
    validatePasswordFromBackend(newPassword);
  };

  const handleConfirmPasswordChange = (e) => {
    const confirmPassword = e.target.value;
    setConfirmPasswordValue(confirmPassword);
    setPasswordMatchValue(confirmPassword === newPasswordInput);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (newPasswordInput === '') {
      setEmptyFieldError(intl.formatMessage(messages['reset.password.empty.new.password.field.error']));
      return;
    }
    if (newPasswordInput !== confirmPasswordInput) {
      setPasswordMatchValue(false);
      return;
    }
    if (passwordValid && passwordMatch) {
      const formPayload = {
        new_password1: newPasswordInput,
        new_password2: confirmPasswordInput,
      };
      props.resetPassword(formPayload, props.token, params);
    }
    mainRef.current.focus();
  };

  if (props.token_status === 'pending') {
    const { token } = props.match.params;
    if (token) {
      props.validateToken(token);
      return <Spinner />;
    }
  } else if (props.token_status === 'invalid') {
    return <InvalidTokenMessage />;
  } else if (props.status === 'success') {
    return <ResetSuccessMessage />;
  } else {
    return (
      <>
        <div id="main" className="d-flex justify-content-center m-4" ref={mainRef}>
          <div className="d-flex flex-column mw-500">
            {props.status === 'failure' ? <Alert variant="danger">{props.errors}</Alert> : null}
            {emptyFieldError && (
              <Alert id="validation-errors" variant="danger">
                <Alert.Heading>{intl.formatMessage(messages['forgot.password.empty.new.password.error.heading'])}</Alert.Heading>
                <ul><li>{intl.formatMessage(messages['reset.password.empty.new.password.field.error'])}</li></ul>
              </Alert>
            )}
            <Form>
              <h3 className="mt-3">
                {intl.formatMessage(messages['reset.password.page.heading'])}
              </h3>
              <p className="mb-4">
                {intl.formatMessage(messages['reset.password.page.instructions'])}
              </p>
              <ValidationFormGroup
                for="reset-password-input"
                invalid={!passwordValid || emptyFieldError !== ''}
                invalidMessage={validationMessage || emptyFieldError}
                className="w-100"
              >
                <Form.Label htmlFor="reset-password-input" className="h6 mr-1">
                  {intl.formatMessage(messages['reset.password.page.new.field.label'])}
                </Form.Label>
                <Input
                  name="new-password1"
                  id="reset-password-input"
                  type="password"
                  placeholder=""
                  onChange={() => setEmptyFieldError('')}
                  onBlur={e => handleNewPasswordChange(e)}
                />
              </ValidationFormGroup>
              <ValidationFormGroup
                for="confirm-password-input"
                invalid={!passwordMatch}
                invalidMessage={intl.formatMessage(messages['reset.password.page.invalid.match.message'])}
                className="w-100"
              >
                <Form.Label htmlFor="confirm-password-input" className="h6 mr-1">
                  {intl.formatMessage(messages['reset.password.page.confirm.field.label'])}
                </Form.Label>
                <Input
                  name="new-password2"
                  id="confirm-password-input"
                  type="password"
                  placeholder=""
                  value={confirmPasswordInput}
                  onChange={e => handleConfirmPasswordChange(e)}
                />
              </ValidationFormGroup>
              <StatefulButton
                type="submit"
                className="btn-primary"
                state={props.status}
                labels={{
                  default: intl.formatMessage(messages['reset.password.page.submit.button']),
                }}
                icons={{ pending: <FontAwesomeIcon icon={faSpinner} spin /> }}
                onClick={e => handleSubmit(e)}
              />
            </Form>
          </div>
        </div>
      </>
    );
  }
  return null;
};

ResetPasswordPage.defaultProps = {
  status: null,
  token_status: null,
  token: null,
  match: null,
  errors: null,
};

ResetPasswordPage.propTypes = {
  intl: intlShape.isRequired,
  resetPassword: PropTypes.func.isRequired,
  validateToken: PropTypes.func.isRequired,
  token_status: PropTypes.string,
  token: PropTypes.string,
  match: PropTypes.shape({
    params: PropTypes.shape({
      token: PropTypes.string,
    }),
  }),
  status: PropTypes.string,
  errors: PropTypes.string,
};

export default connect(
  resetPasswordResultSelector,
  {
    resetPassword,
    validateToken,
  },
)(injectIntl(ResetPasswordPage));
