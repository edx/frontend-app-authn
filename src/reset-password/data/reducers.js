import { RESET_PASSWORD, VALIDATE_TOKEN } from './actions';

export const defaultState = {
  status: null,
  token_status: 'pending',
  token: null,
  errors: null,
};

const reducer = (state = defaultState, action = null) => {
  switch (action.type) {
    case VALIDATE_TOKEN.BEGIN:
      return {
        ...state,
        token_status: 'pending',
      };
    case VALIDATE_TOKEN.SUCCESS:
      return {
        ...state,
        token_status: 'valid',
        token: action.payload.token,
      };
    case VALIDATE_TOKEN.FAILURE:
      return {
        ...state,
        token_status: 'invalid',
        errorCode: action.payload.errorCode,
      };
    case RESET_PASSWORD.BEGIN:
      return {
        ...state,
        status: 'pending',
      };
    case RESET_PASSWORD.SUCCESS:
      return {
        ...state,
        status: 'success',
      };
    case RESET_PASSWORD.FAILURE:
      return {
        ...state,
        status: 'failure',
        errors: action.payload.errors,
        errorCode: action.payload.errorCode,
      };
    default:
      return state;
  }
};

export default reducer;
