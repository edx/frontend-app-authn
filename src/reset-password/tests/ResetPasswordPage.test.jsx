import React from 'react';
import { Provider } from 'react-redux';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { IntlProvider, injectIntl } from '@edx/frontend-platform/i18n';
import * as auth from '@edx/frontend-platform/auth';
import { resetPassword } from '../data/actions';

import ResetPasswordPage from '../ResetPasswordPage';

jest.mock('../data/selectors', () => jest.fn().mockImplementation(() => ({ resetPasswordSelector: () => ({}) })));
jest.mock('@edx/frontend-platform/auth');

const IntlResetPasswordPage = injectIntl(ResetPasswordPage);
const mockStore = configureStore();

describe('ResetPasswordPage', () => {
  let props = {};
  let store = {};

  const reduxWrapper = children => (
    <IntlProvider locale="en">
      <Provider store={store}>{children}</Provider>
    </IntlProvider>
  );

  const submitForm = async (password) => {
    const resetPasswordPage = mount(reduxWrapper(<IntlResetPasswordPage {...props} />));
    await act(async () => {
      resetPasswordPage.find('input#reset-password-input').simulate('blur', { target: { value: password } });
    });
    resetPasswordPage.find('input#confirm-password-input').simulate('change', { target: { value: password } });
    resetPasswordPage.find('button.btn-primary').simulate('click');

    return resetPasswordPage;
  };

  beforeEach(() => {
    store = mockStore();
    props = {
      resetPassword: jest.fn(),
      status: null,
      token_status: 'pending',
      token: null,
      errors: null,
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match reset password default section snapshot', () => {
    props = {
      ...props,
      token: 'token',
      token_status: 'valid',
    };
    const tree = renderer.create(reduxWrapper(<IntlResetPasswordPage {...props} />))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('show spinner component during token validation', () => {
    props = {
      ...props,
      token_status: 'pending',
      match: {
        params: {
          token: 'test-token',
        },
      },
    };
    const tree = renderer.create(reduxWrapper(<IntlResetPasswordPage {...props} />))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match invalid token message section snapshot', () => {
    props = {
      ...props,
      token_status: 'invalid',
    };
    const tree = renderer.create(reduxWrapper(<IntlResetPasswordPage {...props} />))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match pending reset message section snapshot', () => {
    props = {
      ...props,
      token_status: 'valid',
      status: 'pending',
    };
    const tree = renderer.create(reduxWrapper(<IntlResetPasswordPage {...props} />))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match successful reset message section snapshot', () => {
    props = {
      ...props,
      token_status: 'valid',
      status: 'success',
    };
    const tree = renderer.create(reduxWrapper(<IntlResetPasswordPage {...props} />))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match unsuccessful reset message section snapshot', () => {
    props = {
      ...props,
      token_status: 'valid',
      status: 'failure',
      errors: 'Password reset was unsuccessful.',
    };
    const tree = renderer.create(reduxWrapper(<IntlResetPasswordPage {...props} />))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display invalid password message', async () => {
    const validationMessage = 'This password is too short. It must contain at least 8 characters. This password must contain at least 1 number.';
    const data = {
      validation_decisions: {
        password: validationMessage,
      },
    };
    props = {
      ...props,
      token_status: 'valid',
    };

    auth.getHttpClient = jest.fn(() => ({
      post: async () => ({
        data,
        catch: () => {},
      }),
    }));

    const resetPasswordPage = mount(reduxWrapper(<IntlResetPasswordPage {...props} />));
    await act(async () => {
      await resetPasswordPage.find('input#reset-password-input').simulate('blur');
    });
    resetPasswordPage.update();
    expect(resetPasswordPage.find('#reset-password-input-invalid-feedback').text()).toEqual(validationMessage);
  });

  it('with valid inputs resetPassword action is dispatch', async () => {
    const newPassword = 'test-password1';
    props = {
      ...props,
      token_status: 'valid',
      token: 'token',
    };

    auth.getHttpClient = jest.fn(() => ({
      post: async () => ({
        data: {},
        catch: () => {},
      }),
    }));

    const formPayload = {
      new_password1: newPassword,
      new_password2: newPassword,
    };

    store.dispatch = jest.fn(store.dispatch);

    const resetPasswordPage = await submitForm(newPassword);
    expect(store.dispatch).toHaveBeenCalledWith(resetPassword(formPayload, props.token, {}));
    resetPasswordPage.unmount();
  });

  it('should dispatch resetPassword action if validations have reached rate limit', async () => {
    const password = 'test-password';

    auth.getHttpClient = jest.fn(() => ({
      post: async () => {
        throw new Error('error');
      },
    }));
    store.dispatch = jest.fn(store.dispatch);

    props = {
      ...props,
      token_status: 'valid',
      token: 'token',
    };

    const resetPasswordPage = await submitForm(password);
    expect(store.dispatch).toHaveBeenCalledWith(
      resetPassword({ new_password1: password, new_password2: password }, props.token, {}),
    );

    resetPasswordPage.unmount();
  });

  it('should display empty new password field error', () => {
    const validationMessage = 'Please enter your New Password.';
    props = {
      ...props,
      token_status: 'valid',
      token: 'token',
    };

    const resetPasswordPage = mount(reduxWrapper(<IntlResetPasswordPage {...props} />));
    resetPasswordPage.find('button.btn-primary').simulate('click');

    resetPasswordPage.update();
    expect(resetPasswordPage.find('#reset-password-input-invalid-feedback').text()).toEqual(validationMessage);
    expect(resetPasswordPage.find('#validation-errors').first().text()).toEqual('We couldn\'t reset your password.'.concat(validationMessage));
  });
});
