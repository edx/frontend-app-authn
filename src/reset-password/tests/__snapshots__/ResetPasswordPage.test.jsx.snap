// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPasswordPage should match invalid token message section snapshot 1`] = `
<div
  className="d-flex justify-content-center m-4"
  id="main"
>
  <div
    className="d-flex flex-column mw-500"
  >
    <form
      className=""
    >
      <h3
        className="mt-3"
      >
        Reset your password
      </h3>
      <p
        className="mb-4"
      >
        Enter and confirm your new password.
      </p>
      <div
        className="form-group w-100"
      >
        <span />
        <input
          aria-describedby=""
          className="form-control"
          id="reset-password-input"
          name="new-password1"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          placeholder="New password"
          required={true}
          type="password"
          value=""
        />
        <span />
      </div>
      <div
        className="form-group w-100"
      >
        <span />
        <input
          aria-describedby=""
          className="form-control"
          id="confirm-password-input"
          name="new-password2"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          placeholder="Confirm password"
          required={true}
          type="password"
          value=""
        />
        <span />
        <strong
          className="invalid-feedback"
          id="confirm-password-input-invalid-feedback"
        >
          Passwords do not match.
        </strong>
      </div>
      <button
        aria-disabled={false}
        aria-live="assertive"
        className="pgn__stateful-btn pgn__stateful-btn-state-null btn-primary btn btn-primary"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        type="submit"
      >
        <span
          className="d-flex align-items-center justify-content-center"
        >
          Reset my password
        </span>
      </button>
    </form>
  </div>
</div>
`;

exports[`ResetPasswordPage should match pending reset message section snapshot 1`] = `
<div
  className="d-flex justify-content-center m-4"
  id="main"
>
  <div
    className="d-flex flex-column mw-500"
  >
    <form
      className=""
    >
      <h3
        className="mt-3"
      >
        Reset your password
      </h3>
      <p
        className="mb-4"
      >
        Enter and confirm your new password.
      </p>
      <div
        className="form-group w-100"
      >
        <span />
        <input
          aria-describedby=""
          className="form-control"
          id="reset-password-input"
          name="new-password1"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          placeholder="New password"
          required={true}
          type="password"
          value=""
        />
        <span />
      </div>
      <div
        className="form-group w-100"
      >
        <span />
        <input
          aria-describedby=""
          className="form-control"
          id="confirm-password-input"
          name="new-password2"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          placeholder="Confirm password"
          required={true}
          type="password"
          value=""
        />
        <span />
        <strong
          className="invalid-feedback"
          id="confirm-password-input-invalid-feedback"
        >
          Passwords do not match.
        </strong>
      </div>
      <button
        aria-disabled={true}
        aria-live="assertive"
        className="pgn__stateful-btn pgn__stateful-btn-state-pending btn-primary disabled btn btn-primary"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        type="submit"
      >
        <span
          className="d-flex align-items-center justify-content-center"
        >
          <span
            className="pgn__stateful-btn-icon"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-spinner fa-w-16 fa-spin "
              data-icon="spinner"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </span>
          Reset my password
        </span>
      </button>
    </form>
  </div>
</div>
`;

exports[`ResetPasswordPage should match reset password default section snapshot 1`] = `
<div
  className="d-flex justify-content-center m-4"
  id="main"
>
  <div
    className="d-flex flex-column mw-500"
  >
    <form
      className=""
    >
      <h3
        className="mt-3"
      >
        Reset your password
      </h3>
      <p
        className="mb-4"
      >
        Enter and confirm your new password.
      </p>
      <div
        className="form-group w-100"
      >
        <span />
        <input
          aria-describedby=""
          className="form-control"
          id="reset-password-input"
          name="new-password1"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          placeholder="New password"
          required={true}
          type="password"
          value=""
        />
        <span />
      </div>
      <div
        className="form-group w-100"
      >
        <span />
        <input
          aria-describedby=""
          className="form-control"
          id="confirm-password-input"
          name="new-password2"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          placeholder="Confirm password"
          required={true}
          type="password"
          value=""
        />
        <span />
        <strong
          className="invalid-feedback"
          id="confirm-password-input-invalid-feedback"
        >
          Passwords do not match.
        </strong>
      </div>
      <button
        aria-disabled={false}
        aria-live="assertive"
        className="pgn__stateful-btn pgn__stateful-btn-state-null btn-primary btn btn-primary"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        type="submit"
      >
        <span
          className="d-flex align-items-center justify-content-center"
        >
          Reset my password
        </span>
      </button>
    </form>
  </div>
</div>
`;

exports[`ResetPasswordPage should match successful reset message section snapshot 1`] = `
<div
  className="d-flex justify-content-center m-4"
>
  <div
    className="d-flex flex-column"
  >
    <div
      className="text-left mw-500"
    >
      <div
        className="fade alert alert-success show"
        role="alert"
      >
        <div
          className="alert-heading h4"
        >
          Password reset complete.
        </div>
        <span>
          Your password has been reset. 
          <a
            className="font-weight-normal text-info alert-link"
            href="/login"
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            <span>
              Sign in to your account.
            </span>
          </a>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`ResetPasswordPage show spinner component during token validation 1`] = `
<div
  className="d-flex justify-content-center m-4"
  id="main"
>
  <div
    className="d-flex flex-column mw-500"
  >
    <form
      className=""
    >
      <h3
        className="mt-3"
      >
        Reset your password
      </h3>
      <p
        className="mb-4"
      >
        Enter and confirm your new password.
      </p>
      <div
        className="form-group w-100"
      >
        <span />
        <input
          aria-describedby=""
          className="form-control"
          id="reset-password-input"
          name="new-password1"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          placeholder="New password"
          required={true}
          type="password"
          value=""
        />
        <span />
      </div>
      <div
        className="form-group w-100"
      >
        <span />
        <input
          aria-describedby=""
          className="form-control"
          id="confirm-password-input"
          name="new-password2"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          placeholder="Confirm password"
          required={true}
          type="password"
          value=""
        />
        <span />
        <strong
          className="invalid-feedback"
          id="confirm-password-input-invalid-feedback"
        >
          Passwords do not match.
        </strong>
      </div>
      <button
        aria-disabled={false}
        aria-live="assertive"
        className="pgn__stateful-btn pgn__stateful-btn-state-null btn-primary btn btn-primary"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        type="submit"
      >
        <span
          className="d-flex align-items-center justify-content-center"
        >
          Reset my password
        </span>
      </button>
    </form>
  </div>
</div>
`;
